drivers/i3c/i3c_npcx.c

```diff
@@ -0,0 +1,2055 @@
+	}
+#endif /* CONFIG_I3C_USE_IBI */
+
+}
+
+static int npcx_i3c_get_scl_config(struct npcx_i3c_timing_cfg *cfg, uint32_t i3c_src_clk,
+				   uint32_t pp_baudrate_hz, uint32_t od_baudrate_hz)
+{
+	uint32_t i3c_div, freq;
+	uint32_t ppbaud, odbaud;
+	uint32_t pplow_ns, odlow_ns;
+
+	if (cfg == NULL) {
+		LOG_ERR("Freq config NULL");
+		return -EINVAL;
+	}
+
+	if ((pp_baudrate_hz > I3C_SCL_PP_FREQ_MAX) || (od_baudrate_hz > I3C_SCL_OD_FREQ_MAX)) {
+		LOG_ERR("I3C PP_SCL should within 12.5 Mhz, input: %d", pp_baudrate_hz);
+		LOG_ERR("I3C OD_SCL should within 4.17 Mhz, input: %d", od_baudrate_hz);
+		return -EINVAL;
+	}
+
+	/* Fixed PPLOW = 0 to achieve 50% duty cycle */
+	/* pp_freq = ((f_mclkd / 2) / (PPBAUD+1)) */
+	freq = i3c_src_clk / 2UL;
+
+	i3c_div = freq / pp_baudrate_hz;
+	i3c_div = (i3c_div == 0UL) ? 1UL : i3c_div;
+	if (freq / i3c_div > pp_baudrate_hz) {
+		i3c_div++;
+	}
+
+	if (i3c_div > PPBAUD_DIV_MAX) {
+		LOG_ERR("PPBAUD out of range");
+		return -EINVAL;
+	}
+
+	ppbaud = i3c_div - 1UL;
+	freq /= i3c_div;
+
+	/* Check PP low period in spec (should be the same as PPHIGH) */
+	pplow_ns = (uint32_t)(NSEC_PER_SEC / (2UL * freq));
+	if (pplow_ns < I3C_BUS_TLOW_PP_MIN_NS) {
+		LOG_ERR("PPLOW ns out of spec");
+		return -EINVAL;
+	}
+
+	/* Fixed odhpp = 1 configuration */
+	/* odFreq = (2*freq) / (ODBAUD + 2), 1 <= ODBAUD <= 255 */
+	i3c_div = (2UL * freq) / od_baudrate_hz;
+	i3c_div = i3c_div < 2UL ? 2UL : i3c_div;
+	if ((2UL * freq / i3c_div) > od_baudrate_hz) {
+		i3c_div++;
+	}
+
+	odbaud = i3c_div - 2UL;
+	freq = (2UL * freq) / i3c_div; /* For I2C usage in the future */
+
+	/* Check OD low period in spec */
+	odlow_ns = (odbaud + 1UL) * pplow_ns;
+	if (odlow_ns < I3C_BUS_TLOW_OD_MIN_NS) {
+		LOG_ERR("ODBAUD ns out of spec");
+		return -EINVAL;
+	}
+
+	cfg->pplow = 0;
+	cfg->odhpp = 1;
+	cfg->ppbaud = ppbaud;
+	cfg->odbaud = odbaud;
+
+	return 0;
+}
+
+static int npcx_i3c_freq_init(const struct device *dev)
+{
+	const struct npcx_i3c_config *config = dev->config;
+	struct npcx_i3c_data *data = dev->data;
+	struct i3c_reg *inst = config->base;
+	const struct device *const clk_dev = config->clock_dev;
+	struct i3c_config_controller *ctrl_config = &data->common.ctrl_config;
+	uint32_t scl_pp = ctrl_config->scl.i3c;
+	uint32_t scl_od = data->clocks.i3c_od_scl_hz;
+	struct npcx_i3c_timing_cfg timing_cfg;
+	uint32_t mclkd;
+	int ret = 0;
+
+	ret = clock_control_get_rate(clk_dev, (clock_control_subsys_t)&config->clock_subsys,
+				     &mclkd);
+	if (ret != 0x0) {
+		LOG_ERR("Get I3C source clock fail %d", ret);
+		return -EINVAL;
+	}
+
+	LOG_DBG("MCLKD: %d", mclkd);
+	LOG_DBG("SCL_PP_FEQ MAX: %d", I3C_SCL_PP_FREQ_MAX);
+	LOG_DBG("SCL_OD_FEQ MAX: %d", I3C_SCL_OD_FREQ_MAX);
+	LOG_DBG("scl_pp: %d", scl_pp);
+	LOG_DBG("scl_od: %d", scl_od);
+	LOG_DBG("hdr: %d", ctrl_config->supported_hdr);
+
+	/* MCLKD = MCLK / I3C_DIV(1 or 2)
+	 * MCLKD must between 40 mhz to 50 mhz.
+	 */
+	if (mclkd == MHZ(45)) {
```

Comment:
add a define for that 45MHZ (fancy, never realized we had that helper until now)

Reply:
Done